File: day_2/part_1.py
rock = A = X = 1
paper = B = Y = 2
scissors = C = Z = 3

lost = 0
draw = 3
win = 6

input = open('puzzle_input', 'r')
score = 0

for line in input:
    op, me = line.split()
    if op == "A":
        if me == "X":
            score += draw
            score += rock
        elif me == "Y":
            score += win
            score += paper
        elif me == "Z":
            score += lost
            score += scissors
    elif op == "B":
        if me == "X":
            score += lost
            score += rock
        elif me == "Y":
            score += draw
            score += paper
        elif me == "Z":
            score += win
            score += scissors
    elif op == "C":
        if me == "X":
            score += win
            score += rock
        elif me == "Y":
            score += lost
            score += paper
        elif me == "Z":
            score += draw
            score += scissors

print(score)

File: day_2/part_1.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("puzzle_input")
        .expect("This should be able to read in the file");

    let split = input.split("\n");
    let split: Vec<&str> = split.collect();

    let mut sum: i32 = 0;
    for i in &split {
        sum += match *i {
            "" => 0,
            _ => game(&i.chars().nth(0).unwrap(), &i.chars().nth(2).unwrap()),
        }
    }
    println!("{sum}");
}

enum Throw {
    Rock,
    Paper,
    Scissors,
}

enum Outcome {
    Lose,
    Draw,
    Win,
}

fn game(left: &char, right: &char) -> i32 {
    let mut score: i32 = 0;
    let opponent = match left {
        'A' => Some(Throw::Rock),
        'B' => Some(Throw::Paper),
        'C' => Some(Throw::Scissors),
        _ => None,
    };
    let player = match right {
        'X' => Some(Throw::Rock),
        'Y' => Some(Throw::Paper),
        'Z' => Some(Throw::Scissors),
        _ => None,
    };
    let outcome = match (&opponent, &player) {
        (Some(Throw::Rock), Some(Throw::Paper)) => Outcome::Win,
        (Some(Throw::Rock), Some(Throw::Scissors)) => Outcome::Lose,
        (Some(Throw::Paper), Some(Throw::Scissors)) => Outcome::Win,
        (Some(Throw::Paper), Some(Throw::Rock)) => Outcome::Lose,
        (Some(Throw::Scissors), Some(Throw::Rock)) => Outcome::Win,
        (Some(Throw::Scissors), Some(Throw::Paper)) => Outcome::Lose,
        _ => Outcome::Draw,
    };
    score += match &player {
        Some(Throw::Rock) => 1,
        Some(Throw::Paper) => 2,
        Some(Throw::Scissors) => 3,
        _ => 0,
    };
    score += match &outcome {
        Outcome::Win => 6,
        Outcome::Draw => 3,
        Outcome::Lose => 0,
    };
    score
}
